ruler-common/src/main/java/com/spotify/rulercommon/BaseRulerTask.kt

```diff
@@ -0,0 +1,116 @@
+/*
+* Copyright 2023 Spotify AB
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+package com.spotify.rulercommon
+
+import com.spotify.ruler.models.AppFile
+import com.spotify.ruler.models.ComponentType
+import com.spotify.rulercommon.apk.ApkParser
+import com.spotify.rulercommon.apk.ApkSanitizer
+import com.spotify.rulercommon.attribution.Attributor
+import com.spotify.rulercommon.dependency.DependencyComponent
+import com.spotify.rulercommon.models.RulerConfig
+import com.spotify.rulercommon.ownership.OwnershipFileParser
+import com.spotify.rulercommon.ownership.OwnershipInfo
+import com.spotify.rulercommon.report.HtmlReporter
+import com.spotify.rulercommon.report.JsonReporter
+import com.spotify.rulercommon.sanitizer.ClassNameSanitizer
+import com.spotify.rulercommon.sanitizer.ResourceNameSanitizer
+import java.io.File
+
+const val FEATURE_NAME = "base"
+
+interface BaseRulerTask {
```

Comment:
Instead of extending a common base class (or implementing an interface), I feel like it would be cleaner and easier to test if we'd have:
- Something like a `Variant` interface that encapsulates the variant-specific logic. This should be a pure interface without any implementation
- Specific implementations (`GradleVariant`, `CliVariant`, ...) for each variant
- A common `RulerRunner` class that takes in the variant and actually runs the common logic. I.e. something like this:
```kotlin
class RulerRunner {
    fun run(variant: Variant)
}
```
- Each entry point for a variant (`RulerTask` for Gradle, `RulerCli` for CLI, ...) would just set up the variant and pass it to the runner:
```kotlin
class RulerTask {
    fun run() {
        // Maybe some more Gradle-specific logic

        val variant = buildGradleVariant()
        rulerRunner.run(variant)

        // Maybe some more Gradle-specific logic
    }
}
```

That way the code becomes a lot easier to test (you can have a `TestVariant` for unit tests to test behaviour) and easier to expand in the future. Composition is [typically preferrable over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance), even if "multi-inheritance" with Kotlin interfaces is technically possible.

Reply:
Interesting idea, but I think will consider it for later. I think the way we did it now was a smaller change, and is reasonable for what we need to accomplish.